src/packet/bwe/loss_controller.rs

```diff
@@ -0,0 +1,1341 @@
+}
+
+#[derive(Debug, Clone, Copy)]
+struct Observation {
+    num_packets: u64,
+    num_lost_packets: u64,
+    num_received_packets: u64,
+    sending_rate: Bitrate,
+    id: u64,
+    is_initialized: bool,
+}
+
+impl Observation {
+    pub const DUMMY: Self = Self {
+        num_packets: 0,
+        num_lost_packets: 0,
+        num_received_packets: 0,
+        sending_rate: Bitrate::NEG_INFINITY,
+        id: 0,
+        is_initialized: false,
+    };
+
+    pub fn with(partial_observation: &PartialObservation, id: u64) -> Observation {
+        Observation {
+            num_packets: partial_observation.num_packets,
+            num_lost_packets: partial_observation.num_lost_packets,
+            num_received_packets: partial_observation.num_packets
+                - partial_observation.num_lost_packets,
+            sending_rate: Bitrate::from(partial_observation.size as f64 / 1.0),
+            id,
+            is_initialized: true,
+        }
+    }
+}
+
+struct PartialObservation {
+    num_packets: u64,
+    num_lost_packets: u64,
+    size: u64,
+}
+
+impl PartialObservation {
+    pub fn new() -> PartialObservation {
+        PartialObservation {
+            num_packets: 0,
+            num_lost_packets: 0,
+            size: 0,
+        }
+    }
+
+    pub fn update(&mut self, summary: PacketResultsSummary) {
+        // TODO: figure out whether to use usize or u64
+        self.num_packets += summary.num_packets;
+        self.num_lost_packets += summary.num_lost_packets;
+        self.size += summary.total_size;
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+struct ChannelParameters {
+    inherent_loss: f64,
+    loss_limited_bandwidth: Bitrate,
+}
+
+impl ChannelParameters {
+    pub fn new() -> ChannelParameters {
+        ChannelParameters {
+            inherent_loss: 0.0,
+            loss_limited_bandwidth: Bitrate::NEG_INFINITY,
+        }
+    }
+}
+
+trait AsValid<T> {
+    fn as_valid(&self) -> Option<T>;
+    fn is_valid(&self) -> bool;
+}
+
+impl AsValid<Bitrate> for Option<Bitrate> {
+    fn as_valid(&self) -> Option<Bitrate> {
+        if let Some(bitrate) = self {
+            if bitrate.as_f64().is_finite() {
+                return Some(*bitrate);
+            }
+        }
+        None
+    }
+
+    fn is_valid(&self) -> bool {
+        self.as_valid().is_some()
+    }
+}
+
+pub struct LossBasedBweResult {
+    bandwidth_estimate: Option<Bitrate>,
+    state: LossControllerState,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            observation_window_size: 20, // minimum is 2
+            observation_duration_lower_bound: Duration::from_millis(0),
+            trendline_integration_enabled: false,
+            trendline_observations_window_size: 40,
+            temporal_weight_factor: 0.9,
+            instant_upper_bound_temporal_weight_factor: 0.9,
+            instant_upper_bound_loss_offset: 0.05,
+            instant_upper_bound_bandwidth_balance: 75_000.0, // 75 kbps
```

Comment:
can't you use the DataRate type on all bps?

Reply:
Indeed